File: backend/api/v1/endpoints/analysis.py
"""
AI 분석 API 엔드포인트
====================
- 설정 파괴 분석 (비동기 Celery 작업) + DB 캐시
- 분석 작업 상태/결과 조회
"""

from fastapi import APIRouter, HTTPException, status, Depends
from sqlalchemy.orm import Session
from celery.result import AsyncResult
from datetime import datetime

from backend.worker.tasks import detect_inconsistency_task
from backend.worker.celery_app import celery_app
from backend.core.config import settings
from backend.schemas.analysis_schema import ConsistencyRequest
from backend.db.session import get_db
from backend.db.models import Analysis, AnalysisType, AnalysisStatus

router = APIRouter()


# ===== 설정 파괴 분석 - 캐시 조회 =====

@router.get("/consistency/{novel_id}/{chapter_id}")
def get_cached_consistency(novel_id: int, chapter_id: int, db: Session = Depends(get_db)):
    """
    설정 파괴 분석 캐시 조회

    DB에서 가장 최근 COMPLETED 결과를 반환합니다.
    """
    analysis = db.query(Analysis).filter(
        Analysis.novel_id == novel_id,
        Analysis.chapter_id == chapter_id,
        Analysis.analysis_type == AnalysisType.CONSISTENCY,
        Analysis.status == AnalysisStatus.COMPLETED,
    ).order_by(Analysis.completed_at.desc()).first()

    if analysis:
        return {"cached": True, "result": analysis.result}
    return {"cached": False, "result": None}


# ===== 설정 파괴 분석 =====

@router.post("/consistency", status_code=status.HTTP_202_ACCEPTED)
def request_consistency(request: ConsistencyRequest, db: Session = Depends(get_db)):
    """
    설정 파괴 분석 비동기 요청

    Analysis 레코드를 생성하고 Celery 작업으로 비동기 처리합니다.
    """
    # Analysis 레코드 생성 (PENDING)
    analysis = Analysis(
        novel_id=request.novel_id,
        chapter_id=request.chapter_id,
        analysis_type=AnalysisType.CONSISTENCY,
        status=AnalysisStatus.PENDING,
    )
    db.add(analysis)
    db.commit()
    db.refresh(analysis)

    task = detect_inconsistency_task.delay(
        request.novel_id, request.text, request.chapter_id, analysis.id
    )
    return {"task_id": task.id, "status": "PENDING", "analysis_id": analysis.id}


# ===== 분석 작업 결과 조회 =====

@router.get("/task/{task_id}")
async def get_task_result(task_id: str):
    """
    Celery 분석 작업 결과 조회

    Returns:
        status: COMPLETED | FAILED | PROCESSING
        result: 분석 결과 (COMPLETED 시)
        error: 에러 메시지 (FAILED 시)
    """
    result = AsyncResult(task_id, app=celery_app)
    if result.state == 'SUCCESS':
        return {"status": "COMPLETED", "result": result.result}
    elif result.state == 'FAILURE':
        return {"status": "FAILED", "error": str(result.info)}
    return {"status": "PROCESSING"}

File: frontend/src/api/analysis.ts
/**
 * 설정파괴 분석 API 클라이언트
 */

import { request } from './client';

export interface ConsistencyRequest {
    novel_id: number;
    chapter_id?: number;
    text: string;
}

export interface ConsistencyResult {
    status: 'COMPLETED' | 'FAILED' | 'PENDING' | 'PROCESSING';
    result?: {
        status: '설정 파괴 감지' | '설정 일치';
        results: Array<{
            type: '설정 충돌' | '개연성 경고';
            severity?: '치명적' | '주의' | '참고';
            quote: string;
            evidence?: string;
            description: string;
            suggestion: string;
        }>;
    };
    error?: string;
}

export interface ChapterAnalysisRequest {
    novel_id: number;
    chapter_id: number;
    analysis_type: 'plot' | 'style' | 'overall' | 'consistency';
}

export interface PredictionRequest {
    novel_id: number;
    text: string;
}

export interface PredictionResult {
    status: 'COMPLETED';
    result: {
        prediction: string;
    };
}

/**
 * 설정 파괴 분석 캐시 조회
 */
export async function getCachedConsistency(novelId: number, chapterId: number): Promise<{ cached: boolean; result: any }> {
    return request<{ cached: boolean; result: any }>(`/analysis/consistency/${novelId}/${chapterId}`);
}

/**
 * 설정 일관성 검사 요청
 */
export async function requestConsistencyCheck(data: ConsistencyRequest): Promise<{ task_id: string; status: string }> {
    return request<{ task_id: string; status: string }>('/analysis/consistency', {
        method: 'POST',
        body: JSON.stringify(data),
    });
}

/**
 * 작업 결과 조회
 */
export async function getTaskResult(taskId: string): Promise<ConsistencyResult> {
    return request<ConsistencyResult>(`/analysis/task/${taskId}`);
}

/**
 * 스토리 예측 요청
 */
export async function requestPrediction(data: PredictionRequest): Promise<PredictionResult> {
    return request<PredictionResult>('/analysis/prediction', {
        method: 'POST',
        body: JSON.stringify(data),
    });
}

/**
 * 회차 분석 요청 (플롯/문체/종합)
 */
export async function requestChapterAnalysis(data: ChapterAnalysisRequest): Promise<{ task_id: string; status: string; analysis_id: number }> {
    return request<{ task_id: string; status: string; analysis_id: number }>('/analysis/chapter-analysis', {
        method: 'POST',
        body: JSON.stringify(data),
    });
}

/**
 * 회차 분석 캐시 조회
 */
export async function getCachedChapterAnalysis(novelId: number, chapterId: number, analysisType: string): Promise<{ cached: boolean; result: any }> {
    return request<{ cached: boolean; result: any }>(`/analysis/chapter-analysis/${novelId}/${chapterId}/${analysisType}`);
}
